docs/proposals/rescheduling.md

```diff
@@ -0,0 +1,496 @@
+one 4 GB high-priority pod, all of which have sufficient disruption budget, it should
+preempt the two low-priority pods). This is debatable, since all have sufficient
+disruption budget. But still better to err on the side of giving better disruption SLO to
+higher-priority pods when possible?)
+
+Preemption victims must be given their termination grace period. One possible sequence
+of events is
+
+1. The API server binds the preemptor to the node (i.e. sets `nodeName` on the
+preempting pod) and sets `deletionTimestamp` on the victims
+2. Kubelet sees that `deletionTimestamp` has been set on the victims; they enter their
+graceful termination period
+3. Kubelet sees the preempting pod. It runs the admission checks on the new pod
+assuming all pods that are in their graceful termination period are gone and that
+all pods that are in the waiting state (see (4)) are running.
+4. If (3) fails, then the new pod is rejected. If (3) passes, then Kubelet holds the
+new pod in a waiting state, and does not run it until the pod passes passes the
+admission checks using the set of actually running pods.
+
+Note that there are a lot of details to be figured out here; above is just a very
+hand-wavy sketch of one general approach that might work.
+
+See #22212 for additional discussion.
+
+### Node drain
+
+Node drain will be handled by one or more components not described in this document. They
+will respect disruption budgets. Initially, we will just make `kubectl drain`
+respect disruption budgets.  See #17393 for other discussion.
+
+### Rescheduler
+
+All rescheduling other than preemption and node drain will be decided and enacted by a
+new component called the *rescheduler*. It runs continuously in the background, looking
+for opportunities to move pods to better locations. It acts when the degree of
+improvement meets some threshold and is allowed by the pod's disruption budget.  The
+action is eviction of a pod using the `/evict` subresource, with the pod's signature
+enqueued in the node's `PreferAvoidPods`. It does not force the pod to reschedule to any
+particular node. Thus it is really an "unscheduler"; only in combination with the evicted
+pod's scheduler, which schedules the replacement pod, do we get true "rescheduling."  See
+the "Example use cases" section earlier for some example use cases.
+
+The rescheduler is a best-effort service that makes no guarantees about how quickly (or
+whether) it will resolve a suboptimal pod placement.
+
+The first version of the rescheduler will not take into consideration where or whether an
+evicted pod will reschedule. The evicted pod may go pending, consuming one unit of the
+corresponding shard strength disruption budget by one indefinitely. By using the `/evict`
+subresource, the rescheduler ensures that an evicted pod has sufficient budget for the
+evicted pod to go and stay pending.  We expect future versions of the rescheduler may be
```

Comment:
Thinking about it more, I'm not convinced we can really get away with having the rescheduler not know at least the required predicate functions, even for the first version of the rescheduler.

Take for example the "move pods onto under-utilized" nodes use case. It's important to know if an evicted pod will actually not reschedule onto any of those nodes--for example those nodes might be under-utilized because they're being drained and are marked unschedulable, or maybe they have some kind of taint due to being in a dedicated node group or otherwise restricted to a limited set of pods. In such cases the eviction is pointless, as the evicted pod will not move onto any of those nodes. On the other hand, maybe the nodes were just added by cluster auto-scale-up, in which case they will be feasible for the pod you're considering moving. It seems important to be able to distinguish these two cases.

This is a slightly different argument than saying rescheduler needs the predicate functions so that it can know whether an evicted pod will go pending. I think the argument that we don't really care whether the pod goes pending is reasonable, certainly for a first version of the rescheduler. The argument here is that rescheduler needs the predicate functions so it can know if the node(s) that seem better for the pod in question are actually feasible for that pod. If they're not, the pod might not end up on a node that is actually better.